dotnet/src/PF-Extensions/Planning.ChatStepwisePlanner/ChatStepwisePlanner.cs

```diff
@@ -0,0 +1,539 @@
+        }
+
+        promptConfig.Completion.MaxTokens = this.Config.MaxTokens;
+
+        //this._systemStepFunction = this.ImportSemanticFunction(this._kernel, "StepwiseStep", promptTemplate, promptConfig);
+        this._nativeFunctions = this._kernel.ImportSkill(this, RestrictedSkillName);
+
+        this._context = this._kernel.CreateNewContext();
+        this._logger = this._kernel.Log;
+    }
+
+    public Plan CreatePlan(string goal)
+    {
+        if (string.IsNullOrEmpty(goal))
+        {
+            throw new PlanningException(PlanningException.ErrorCodes.InvalidGoal, "The goal specified is empty");
+        }
+
+        string functionDescriptions = this.GetFunctionDescriptions();
+
+        Plan planStep = new(this._nativeFunctions["ExecutePlan"]);
+        planStep.Parameters.Set("functionDescriptions", functionDescriptions);
+        planStep.Parameters.Set("question", goal);
+
+        //planStep.Outputs.Add("agentScratchPad");
+        planStep.Outputs.Add("stepCount");
+        planStep.Outputs.Add("skillCount");
+        planStep.Outputs.Add("stepsTaken");
+
+        Plan plan = new(goal);
+
+        plan.AddSteps(planStep);
+
+        return plan;
+    }
+
+    [SKFunction, SKName("ExecutePlan"), Description("Execute a plan")]
+    public async Task<SKContext> ExecutePlanAsync(
+        [Description("The question to answer")]
+        string question,
+        [Description("List of tool descriptions")]
+        string functionDescriptions,
+        SKContext context)
+    {
+        var stepsTaken = new List<SystemStep>();
+        this._logger?.BeginScope("ChatStepwisePlanner");
+        if (!string.IsNullOrEmpty(question))
+        {
+            this._logger?.LogInformation("Ask: {Question}", question);
+
+            // TODO MOVE
+
+            var systemPromptTemplate = EmbeddedResource.Read("Microsoft.SemanticKernel.Planning.ChatStepwise.Prompts.systemprompt.txt");
+            var fewShotExamplesTemplate = JsonSerializer.Deserialize<FewShotExample[]>(EmbeddedResource.Read("Microsoft.SemanticKernel.Planning.ChatStepwise.Prompts.fewshotexamples.json"));
+
+            // TODO use semantic function to fill in function descriptions
+            string currentTime = DateTimeOffset.UtcNow.ToString("f", DateTimeFormatInfo.InvariantInfo);
+            ChatHistory history = this._chatCompletion.CreateNewChat(systemPromptTemplate.Replace("{{$functionDescriptions}}", functionDescriptions).Replace("{{time.UtcNow}}", currentTime));
+            foreach (FewShotExample example in fewShotExamplesTemplate)
+            {
+                switch(example.Role)
+                {
+                    case Role.User:
+                        history.AddUserMessage(example.Content);
+                        break;
+
+                    case Role.Assistant:
+                        history.AddAssistantMessage(example.Content);
+                        break;
+                }
```

Comment:
>} [](http://example.com/codeflow?start=16&length=1)

1. default should throw exception
2. slightly newer syntax might be shorter

something like:
example.Role switch
{
    Role.User => example.Content,
    Role.Assistant => example.Content,
    _ => throw new NotSupportedException(...)
});